src/ApiService/ApiService/onefuzzlib/VmssOperations.cs

```diff
@@ -195,26 +198,40 @@ private Task<string> GetInstanceIdForVmId(Guid scaleset, Guid vmId)
         => _cache.GetOrCreateAsync(new InstanceIdKey(scaleset, vmId), async entry => {
             var scalesetResource = GetVmssResource(scaleset);
             var vmIdString = vmId.ToString();
-            await foreach (var vm in scalesetResource.GetVirtualMachineScaleSetVms().AsAsyncEnumerable()) {
-                var response = await vm.GetAsync();
-                var instanceId = response.Value.Data.InstanceId;
-                if (response.Value.Data.VmId == vmIdString) {
-                    // we found the VM we are looking for
-                    entry.AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(10);
-                    return instanceId;
-                } else {
-                    // if we find any other VMs, put them in the cache
-                    if (Guid.TryParse(response.Value.Data.VmId, out var vmId)) {
-                        using var e = _cache.CreateEntry(new InstanceIdKey(scaleset, vmId));
-                        _ = e.SetValue(instanceId);
-                        e.AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(10);
+
+            var node = await _nodeOps.GetByMachineId(vmId);
+            var existsInTable = node is not null && node.InstanceId is not null;
+            if (existsInTable) {
+                entry.AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(10);
+                _ = entry.SetValue(node!.InstanceId!);
+                return node.InstanceId!;
+            } else {
+                await foreach (var vm in scalesetResource.GetVirtualMachineScaleSetVms().AsAsyncEnumerable()) {
+                    var response = await vm.GetAsync();
+                    var instanceId = response.Value.Data.InstanceId;
+
+                    _ = await _nodeOps.SetInstanceId(scaleset, vmId, instanceId);
+
+                    if (response.Value.Data.VmId == vmIdString) {
+                        // we found the VM we are looking for
+                        entry.AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(10);
+                        _ = entry.SetValue(instanceId);
```

Comment:
I added "entry.SetValue" - since we should cache instance id of the found node. Am I wrong ?

Reply:
The return value is used to set the value. of the entry when we call GetOrCreateAsync so no need to use SetValue here. 
This can actually be dangerous if the implementation does not allow reentrancy on the same entry

Reply:
Note that we use setValue below for other entries in the cache not the one that is currently open.

Reply:
reverted